m9 -> m10

hayesgm opened the pull request:
This patch includes the previous m9 changes as a new m10 branch rebased off of m9, which would be force-pushed to develop.

hayesgm committed:
Add Gateway REPL

This patch adds a Gateway repl to make it easy to interact with a Gateway instance. For example, the following examples:

```sh
yarn repl -c testnet
```

Query the current cash yield

```sh
> (await api.query.cash.cashYield()).toJSON()
300
```

Show the current test-net block

```sh
> .block
```

Show the current test-net validators

```sh
> .validators
  5CiPPseXPECbkjWCa6MnjNokrgYjMqmKndv2rSnekmSK2DjL:
    substrate_id=0x1cbd2d43530a44705ad088af313e18f80b53ef16b36177cd4b77b846f2a5f07c
    eth_address=0x286f2a10c28c966e97a72b8246041fbf636e673e

  5HMqNs9offzpiebeuzHVwKcJRid1L44KgEKF7jwpYYFM25kY:
    substrate_id=0xea3da7e5b24ee22ce5fa252136745bfbefcb657201404f4479bcbe42135e234c
    eth_address=0x4515e1ce5d4c42da4b0561f52ef12dee19f9c020

  5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty:
    substrate_id=0x8eaf04151687736326c9fea17e25fc5287613693c912909cb226aa4794f26a48
    eth_address=0xc9b0c3ed4efa833a7ad5459755a18f9689a0f7ac

  5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY:
    substrate_id=0xd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d
    eth_address=0x55413a2d4908d130c908ccf2f298b235bacd427a
```

Show the Starport admin:

```sh
> await starport.methods.admin().call()
'0x046231a12d30248bAD3322Af74CEA9c325627D32'
```

hayesgm committed:
Show the Dangers of a Re-org

This patch adds a test that currently shows the danger of our a re-org. We will use this as one of the bases to test that our new work on re-org resistance is effective in real-life scenarios.

hayesgm committed:
Add CrazyLock Scenario

waynenilsen committed:
CLI Configuration (#264)

* also patch ETH_KEY_ID situation
* update integration tests to use CLI configuration

jflatow committed:
Some tiny cleanups (#260)

hayesgm committed:
Repl updates

hayesgm committed:
Add better handling when missing corresponding types.json

hayesgm committed:
Deploy New Starport

This patch deploys a new Starport with the Timelock as the admin, which is the current Starport Impl now used for test-net. We add a script to make it fairly easy to deploy a new Starport implementation.

hayesgm committed:
Cleanup and add TODOs for Re-org

maxwolff committed:
First Benchmarks (#268)

* stub

* Progress debugging

* add first weights benchmark

* receive event debug

* publish sig

Co-authored-by: Jared Flatow <jflatow@gmail.com>

toninorair committed:
validate unsigned + exec_trx_request tests (#266)

hayesgm committed:
Derive Types

This patch adds a set of macros to derive types.json automatically. This is, all in all, a little complicated because the best way to analyze the code is through macros (since we can easily mark types that need to be exported) but macros, for obvious reasons, are run at compile-time and are not _meant_ to be side-effecty. That is, you can't easily just store a bunch of global data and when you're finished just emit it. Instead, we truncate and re-write a file-system temp file each time we encounter a new type and eventually we stop encountering new ones and we're good. We do this separately for each project in `scripts/build_types.json` since partial compilation and project-by-project compilation makes this even harder to do. We also need to clean and rebuild each time to ensure every macro is run. _Also_ there is some compile time slowness added here since I can't seem to remove the macros via features or other settings since this runs in the same layer as those types of tools. All-in-all, it's significantly fast enough and hopefully will make the process of updating our types.json trivial going forward.

Note: some of the code is pretty spaghetti in the derive code itself and can be upgraded over time. The issue is that there is a ton of possibilities that could occur in the syntax and we probably don't want to handle all of them.

The most complex one that we need to consider is `WIDTH` which, as a const, is not very easy to determine its value, so for now, we just use its value directly, since literals are fairly easy.

Note: the goal of `types-derive` is that it is generic and can be added to substrate's core, so we try to avoid "Gateway-specific" features in the code.

hayesgm committed:
Ensure new types seems to match old types as much as possible

hayesgm committed:
Fix more types to avoid [] syntax and add types for Chains

jflatow committed:
Start documenting release process (#273)

toninorair committed:
Add logging for validate_unsigned (#277)

coburncoburn committed:
Cob/account liquidity experiment (#279)

* account liqudity api experiment

* add gateway namespace, add portfolio endpoint

* playbook to update binary on full nodes ( update rpc )

* add exec trx request to repl

coburncoburn committed:
Cob/sequential trx requests (#278)

* add sorting of trx requests in transaction pool

format

re-remove split_once directive

* fix test

* fix test

hayesgm committed:
Start to Improve Scenarios as Regression Tests (#275)

* Start to Improve Scenarios as Regression Tests

This patch starts to improve scenarios so that they can pass consistently. We make a few changes:

1. We add a `QUIET_SCENARIOS` flag that pushes all scenario output to temp files. This makes it possible to run scenarios and see the output and not 1,000 pages of validator text.
2. We add an `EventTracker` module that handles tracking events without using global vars (which was probably messing up test cases). Events are hard because they come in from a subscription in PolkadotJS, but our scenarios often ask "Have you seen any Substrate `Lock()` events?" and we have to say something like "Yes, we saw one more recently than the last time you asked." That can be a bit tricky and now that state is encapsulated and doesn't leak from one scenario to the next (this may have been a large bug in running sequential tests).
3. We start to move back away from `--runInBand` which means tests will run in separate processes and concurrently again. We may want to make it easier to not run in band for building tests (it's just a matter of passing the flag, but engineers might forget/not know how to do this, so we should make that clear).
4. We remove `only: true` and start to add back multiple tests so we can measure what is working and not. So far the results are not too excessively shabby.

* Continue to fix integration tests

* Add RPC decoration to preclude warnings

* Make scenarios quiet in CI

* Fix line number errors

* Track sleeps for better teardown procedure

* Try to quiet them down

* Increase timeout

* Try again

* Get more tests passing

* Get last upgrade scen working again

* Finish prices tests

* Add complex scenario for borrow cash interest

* Add features for integration mode

* More cleanup and fixes

* Clean up Gov tests

* Clean up more tests, including prices

* Fix some types issues and merge conflicts

* It was always a String, let's make it so

hayesgm committed:
Liquidation Scenarios (#280)

* Liquidation Scenarios

This patch starts to build a full array of liquidation scenarios and starts to uncover some buggy behavior. There are still more tests to write, but we'll need to fix the specification in regards to close factor limitations and healthiness checks.

* Add final liquidation tests

hayesgm committed:
Max Cash Less Fee (#283)

* Max Cash Less Fee

This patch fixes "Transfer Max Cash" to subtract out the fee before calculating "max" so it doesn't necessarily push you into a negative cash position (since, e.g. if you had 5 CASH and then you transfered max, we would calculate that as "you want to transfer 5 Cash" and then charge you 0.01 Cash as a transfer fee and end up with a -0.01 CASH balance. Now we say you want to transfer 4.99 and you'll end up with exactly 0 CASH balance afterwards).

Additionally, we un-mock the tests and start to build a nice little suite of unit tests. Currently liqudation tests aren't implemented because I got tired, but the rest are starting to look good.

* Get significantly higher test coverage up for exec_trx_request

* Fix transfer max; add partial debug compilation

hayesgm committed:
Liquidate Liquidity Check (#284)

This patch adds a liquidity check to a user to ensure they are underwater before we allow liquidations on the account. There are still other areas to consider (i.e. close factor and collateral quantity), but this is a first start on fixing some bugs in the liquidation function.

Additionally, we add significant test coverage to the `has_liquidity_xxx` functions to ensure they are all individually test in a range of scenarios. We found no serious bugs in that process, but we should consider breaking on some invalid inputs (e.g. for unsupported assets, etc). These are currently left as TODOs.

hayesgm committed:
Handle Unpriced and Mismatched Assets (#285)

* Handle Unpriced and Mismatched Assets

To improve completeness, we add checks for matching asset prices to the assets themselves and ensure prices are set when checking liquidity. We change the behavior from pallet-oracle to return an optional price and use a default zero in the places where it's currently expected. We make no effort here to decide if _that_ is the correct behavior- just to keep the status quo in that regard. A few other call sites may have assumed zero prices that will now receive an error, but that is _probably_ a good thing.

* Remove outdated comments

jflatow committed:
Event processing should handle reorgs / undo (#287)

This still needs more tests and to implement undo lock functionality.
It especially needs to be tested that events are applied correctly after ingestion.
However, it now compiles / passes existing tests, and could be merged to reduce future conflicts.

hayesgm committed:
Cash Pipeline (#288)

* Factor out `Liquidate.rs` and `BalanceHelpers.rs` with Unit Tests

This patch starts to factor out `liquidate.rs` and `balance_helpers.rs` to their own modules in `internal/`. We significantly increase the test coverage for these functions, finding a few minor cases that could have otherwise panic'd. We thus return `results` here for these functions.

I intend to add unit tests to `liquidate.rs`, but I needed the helper functions to be shared first, which was completed in the current patch.

* Vec variant

* Move to BTree for caching

* Start to add portfolio, as well

* With fees

* Get liquidate and transfers working on effectful

* Get unit tests passing

* Reduce cloning and change name to check_collateralized

* Remove unused liquidity checks

hayesgm committed:
Limit Liqudations by Spec (#291)

* Limit Liqudations by Spec

This patch fixes liqudiations by limiting when they can occur, specifically prohibiting you from liquidating a borrow into a positive position (close > 100%) and preventing you from liquidating collateral into a borrow position (seize > 100% collateral). These are important fixes for real-life.

We will add better unit tests, but we are waiting on fixes for scenarios to test the current scenarios for this behavior.

* Add coverage for  and

* Add significantly more liquidate tests

jflatow committed:
Assorted Patches (#290)

* Make the handling of liquidate clearer / consistent with others

* Fix up signature recognition for notices

* Add eth_starport_parent_block to chain_spec properties

* Fix liquidate tests given cannot force borrower into collateral debt

hayesgm committed:
Break Transfer into Constituent Actions (#292)

* Break Transfer into Constituent Actions

* Start to get chain cash working

* Use old reason

* Finish migration to Cash Pipeline for remaining functions and clear out now unused functions from core

* fix test

* Add block info to generated chain spec

* Remove unused functions

jflatow committed:
Use last block property from runtime interface as default (#294)

hayesgm committed:
Remove outdated Portfolio::from_storage fn (#296)

We now just use `CashPipeline::load_portfolio` which has the same behavior (but can be shared with pipelines). There must have been a disconnect with from storage since it was calling back into `core::` code that was being upgraded to deal with pipelines. Once we moved everything onto the pipeline code, everything works. There's some awkwardness to `Portfolio`s being generated by `pipeline.rs` now, but since they basically co-depend on each other, we'll probably need to make a new common ancestor module to reconcile a better API pattern. Overall this doesn't seem worth it at the moment since everything works now and we have effectively no code duplication left.

hayesgm committed:
Add Unit Tests for `assets.rs` and `transfer.rs` (#295)

This patch simply adds unit tests for `assets.rs` and `transfer.rs` to make sure we actually have a large amounts of unit test coverage for our core functionality. Nothing surprising was found during these tests (except continued questions on the "right" value for TotalCashPrincipal and ChainCashPrincipals).

maxwolff committed:
Add weights for every extrinsic (#281)

* stub

* progress

* fmt

* cleanups

* fix min tx val

* fix events rebase

* fmt

* fix liquidate

* remove hex

* polish

maxwolff committed:
fix test (#299)

hayesgm committed:
A ton of minor cleanups and refactorings (#298)

Nothing here should change any behavior, but I've started to move the rest of the code out of `core.rs` and into more correspondingly-typed modules. E.g. the asset fetching code has moved to `assets.rs` and the on initialize code to `initialize.rs`. This will open up the door to better unit testing as it's more clear what is and is not being covered by unit tests right now. It's important to merge this quickly as it touches a lot of random lines of code.

jflatow committed:
Fix initialization of starport parent block (#297)

* Fix initialization of starport parent block

* Scen debugging tweaks

maxwolff committed:
Change auth requires supermajority (#300)

* change auth supermajority

* fmt

* camelcase

* add comment

coburncoburn committed:
add more data about cash to the cashdata rpc (#289)

* add more data about cash to the cashdata rpc

* delete unused params

* fix rebase

coburncoburn committed:
add trx request to the and_provides to fix stuck nonces on failed trx… (#301)

* add trx request to the and_provides to fix stuck nonces on failed trx requests

* fix test

jflatow committed:
Resolve some open questions / standing issues in rates (#303)

jflatow committed:
Minimum blocks wait for all events (#304)

jflatow committed:
Fix substrate coverage (#305)

jflatow committed:
Add rudimentary filter on worker side of already signed blocks (#307)

* Add rudimentary filter on worker side of already signed blocks

Still need to determine worker's identity before this can take effect.

* Add ability for workers to get current validator, unlock filtering

jflatow committed:
Event Fixes (#309)

* Start to test scenarios

* Debugging improvements

* Better handling new Web3 revert without reasons

* Fix test failures due to more blocks being fetched

* Add missing check for already signed reorg (also fixes warnings)

Co-authored-by: Geoffrey Hayes <hayesgm@gmail.com>

jflatow committed:
Debugging Reorgs (#311)

* Debugging Reorgs

* Some cleanup / notes

Co-authored-by: Geoffrey Hayes <hayesgm@gmail.com>

hayesgm committed:
Debugging Reorgs (#310)

* Fix Cash RPC, Add Max Ingression Blocks, and Misc. Scenario Fixes

This patch fixes the Cash RPC which was double adding interest to the cash results. We've also added a max ingression blocks since certain scenario tests attempted to lock unpriced assets and thus failed. I would debate after a sufficient number of blocks, we should always just take in a given event from the queue (e.g. 50 or 100 blocks). Finally, we fix a large number of scenarios, including making sure we use `types.json` from earlier versions for upgrade tests. We are still facing a number of issues testing upgrades, and thus those are marked as skipped for now. Finally, we switch back from 1-second blocks to auto-mine for the majority of test cases. This is because they run about 50% faster when auto-mining. To handle the ingression queue code, we just mine 100 blocks whenever we're waiting for Eth events.

* Resolve some XXXs

Co-authored-by: Jared Flatow <jflatow@gmail.com>

toninorair committed:
Add checks for Extract, Liquidate and Transfer events (#312)

hayesgm committed:
Fix m8 Failure (#313)

This patch fixes failures running m8 due to the enum shifting between versions. We add a reserved filler into the enums that can later be set to a different value, e.g. for another chain. We try to fill all of the slots since we would prefer that the numbering of chains stays constant and doesn't shift (e.g. it's #0 for ChainAccounts but #1 for ChainAssets and #5 for ChainBlocks).

Currently we cannot run m7 wasm in the native shell. This is likely due to some incompatability in the host functions versus the previously built wasm code. This new test shows off that behavior (it was broken with a Wasmi Trap). We fix the issue by changing the Dot/Eth order in chains.rs (not sure why this is the fix, but it worked). Note: this test now traps on missing a host function for getting cash data, but this is **not of any concern**. The original error was that blocks could not form, and we have a partial fix here that warrants more follow up. That is: why did this fix work? How can we make it work without changing the enum orders?

Again, the current error in the test is not the issue. The issue is "fixed" as we can mine blocks, but we need to figure out why this fix worked and if we can fix it without the "hack".

jflatow committed:
Miner cumulative, asset modified events, and some other fixes (#315)

* Miner cumulative, asset modified events, and some other fixes

* MinerPaid events; initialize events; appease the tests

maxwolff committed:
Unit test reorg (#317)

* unit test reorg

* fmt

* fmt

* fix rebase

toninorair committed:
Fix substrate tests ? (#316)

maxwolff committed:
Run benchmarks on digital ocean instance (#276)

Roughly matches the original substrate benchmarking hardware.

toninorair committed:
Move deadline to const (#318)

coburncoburn committed:
bump spec version (#314)

hayesgm committed:
Start to get upgrade to m9 test going (#321)

This patch gets the upgrade tests working. They are pretty crazy due to the stateful API connections that make it quite difficult to send an RPC that changes the encoding/decoding scheme, since they can't really be on version 1 (since they crash after changeover as soon as they see an event) nor version 2 (since they crash trying to init with an old types.json). You can try to blend the types.json from both versions (we do this) but you'll still get truly conflicting decoding schemes. It's just a mess so we actually mostly try to disconnect the APIs and reconnect them after the change. It's a bit flakey but it works somewhat okay.

With this strategy, we fix the previous upgrade scenarios and we add an m8 -> m9 scenario.

We were unable to get into the state seen on test-net (not sure how to repro), and thus we still don't have full confidence this upgrade will fix the issue, but we see no reason that it should specifically break it, either.

hayesgm committed:
Native m8 Shell (#324)

This patch adds a scenario where we run a native m8 binary and upgrade _it_ to m9 (as opposed to running an m9 binary with m8 WASM runtime and then do the upgrade there). This can detect anomolies in the node runtime that wouldn't be found by other tests. This test _is passing_ but since it requires a binary build in the release directory that is currently hand-built, it's skipped in CI.

hayesgm committed:
Start to add real upgrade path to m9

jflatow committed:
Const compile config

hayesgm committed:
Start to get replacements working

hayesgm committed:
The hack of the century

jflatow committed:
Set testnet params and delete logs

hayesgm committed:
Add support for testnet to releases

jflatow committed:
WIP: Add stubnet feature flag

hayesgm committed:
More features

hayesgm committed:
Get scenario test truly passing

hayesgm committed:
Finish setting up release code

hayesgm committed:
Minor test fixes

hayesgm committed:
More aspects of m9

hayesgm committed:
Fix validation code

hayesgm committed:
Bump to m10

jflatow commented on chains/scripts/add_users.js:
```diff
@@ -0,0 +1,50 @@
+async function loadem(starting_page = 0) {
```
Hm what is this file actually?

hayesgm replied:
I think Coburn was using this for... load testing?

jflatow reviewed (approved):
I almost forgot how huge this release is. Amazing work 🙌  Let's 🚢 before we change our minds!

hayesgm commented on .github/workflows/test-workflow.yml:
```diff
@@ -40,7 +41,7 @@ jobs:
         run: |
           set -eo pipefail
-          cargo test -- -Z unstable-options --format json | tee unit-test-results.json
+          cargo test -- -Z unstable-options --format json --test-threads 1 | tee unit-test-results.json
```
Note: we shouldn't _need_ this.

hayesgm commented on .github/workflows/test-workflow.yml:
```diff
@@ -237,14 +238,14 @@ jobs:
       - name: Build Gateway
         run: |
-          cargo build
+          cargo build --release --features freeze-time --features runtime-debug
```
Note: we should probably use `--features integration` which properly includes sub-features.

hayesgm replied:
Addressing in my tweaks branch.

hayesgm commented on README.md:
````diff
@@ -219,4 +219,47 @@ by appending your own. A few useful ones are as follow.
+Attach the following files to the release, from the repository where you built the release artifacts:
+
+```
+contracts.json
````
Note: these are lightly out of date.

hayesgm commented on chains/build_spec.js:
```diff
@@ -97,6 +97,7 @@ async function setAuthorities(chainSpec, chainConfig, opts) {
 async function setStarport(chainSpec, chainConfig, opts) {
   let chainDeployment = await fetchChainDeployment(chainConfig.eth_network);
   let starportAddress = must(chainDeployment.Contracts, 'Starport');
+  // TODO: Set other properties
```
TODO: We should probably remove properties.

waynenilsen replied:
I'm pretty sure the `properties` key is buried somewhere in substrate but we can remove all of the `properties` that we looked for.

hayesgm commented on chains/testnet/chain-spec-raw.json:
```diff
@@ -6,7 +6,12 @@
   "protocolId": "local",
   "properties": {
-    "eth_starport_address": "0xD905AbBA1C5Ea48c0598bE9F3f8ae31290B58613"
+    "eth_starport_address": "0xD905AbBA1C5Ea48c0598bE9F3f8ae31290B58613",
```
TODO: Remove properties?

hayesgm commented on chains/testnet/chain-spec.json:
```diff
@@ -6,7 +6,12 @@
   "protocolId": "local",
   "properties": {
-    "eth_starport_address": "0xD905AbBA1C5Ea48c0598bE9F3f8ae31290B58613"
+    "eth_starport_address": "0xD905AbBA1C5Ea48c0598bE9F3f8ae31290B58613",
```
TODO: Remove properties?

hayesgm commented on ethereum-client/src/lib.rs:
```diff
@@ -126,96 +199,227 @@ fn send_rpc(
+    let block_obj = get_block_object(server, &block_str)?;
+    debug!("eth_starport_address: {:X?}", &eth_starport_address[..]);
+    let get_logs_params = vec![serde_json::json!({
+        "address": format!("0x{}", &eth_starport_address[..].iter().map(|x| format!("{:02x}", x)).collect::<String>()),
```
We should remove this hack and just use `::hex::encode` now that we don't do WASM replacements.

hayesgm commented on ethereum-client/src/lib.rs:
```diff
@@ -126,96 +199,227 @@ fn send_rpc(
+pub fn get_latest_block_number(server: &str) -> Result<u64, EthereumClientError> {
+    let response_str: String = send_rpc(server, "eth_blockNumber".into(), vec![])?;
+    let response = deserialize_block_number_response(&response_str)?;
+    debug!("eth_blockNumber response: {:?}", response.result.clone());
```
Not sure this debug statement is actually useful?

hayesgm commented on integration/__tests__/borrow_scen.js:
```diff
@@ -22,13 +21,15 @@ async function lockZRX({ ashley, bert, bat, zrx }) {
   {
     name: "Borrow BAT and Garner Interest",
-    scenario: async ({ ashley, bert, bat, zrx, chain, starport, log, cash }) => {
+    notes: "This test allows arbitrary time passage, and thus has some estimation",
```
TODO: This should really just use freeze time, now.

hayesgm commented on integration/__tests__/gov_scen.js:
```diff
@@ -21,183 +19,46 @@ buildScenarios('Gov Scenarios', gov_scen_info, [
       expect(await chain.nextCodeHash()).toEqual(currHash);
 
-      let event = await chain.setNextCode(await curr.wasm());
+      let event = await chain.setNextCode(await curr.wasm(), curr, false);
```
Are these params right?

hayesgm replied:
They are.

hayesgm commented on integration/__tests__/liquidate_scen.js:
```diff
@@ -1,46 +1,337 @@
+      `A simple replay of the collateral-for-collateral liquidation scenario, but Ashley
+       doesn't have much of the seized collateral (e.g. she's mostly in USDC, not ETH).
+
+       Note: This test is a counter-factual and should *not* pass`,
```
This... message should be removed now, right?

hayesgm commented on integration/__tests__/lock_scen.js:
```diff
@@ -109,22 +107,22 @@ buildScenarios('Lock Scenarios', lock_scen_info, [
       });
       let data = getEventData(event);
-      expect(data.CashPrincipalAmount).toBeCloseTo(99999996);
+      expect(data.CashPrincipalAmount).toBeCloseTo(99999996, -1); // TODO: Check this better
```
TODO: Use freeze-time?

hayesgm commented on integration/__tests__/upgrade_to_m9_m10_scen.js:
```diff
@@ -0,0 +1,158 @@
+    }
+  },
+  {
+    only: true, // Currently CI doesnt have native binaries
```
only -> skip?

hayesgm commented on integration/util/scenario/chain.js:
```diff
@@ -197,16 +227,21 @@ class Chain {
   }
 
   async cullNotices() {
```
TODO: Can we remove this now? We might need to remove earlier integration tests that depended on this function, but that's probably okay?

hayesgm commented on integration/util/scenario/chain.js:
```diff
@@ -254,50 +289,57 @@ class Chain {
+    let argsEncoded = zip(argTypes, argValues).map(([argType, argValue]) => {
+      return registry.createType(argType, argValue).toHex().slice(2);
+    });
+    return `0x04${hexByte(palletNumber)}${hexByte(callNumber)}${argsEncoded.join('')}`;
```
TODO: This probably should _not_ start with 0x04 (Extrinsic Version)

hayesgm commented on integration/util/scenario/chain_spec.js:
```diff
@@ -48,43 +48,49 @@ async function baseChainSpec(validatorsInfoHash, tokensInfoHash, ctx) {
       }
     },
-    miner_shares: 1000,
+    miner_shares: 0,
```
TODO: Do we want miner shares for testing? Do we test that, generally?

hayesgm commented on integration/util/scenario/chain_spec.js:
```diff
@@ -48,43 +48,49 @@ async function baseChainSpec(validatorsInfoHash, tokensInfoHash, ctx) {
   let reporters = ctx.__reporters();
 
+  let cashGenesis = {};
+  if (genesisVersion.supports('eth-starport-parent-block')) {
```
This is useful.

hayesgm commented on integration/util/scenario/ctx.js:
```diff
@@ -34,10 +37,24 @@ class Ctx {
 
+  __initialYieldStartMS() {
+    let raw = this.__initialYieldStartRaw();
+    if (raw > 4102444800) { // Jan 1, 2100 as seconds since 1970
```
Basically, we're just trying to see if the time is so large that it _must_ have been ms and not s

hayesgm commented on integration/util/scenario/validator.js:
```diff
@@ -166,9 +210,11 @@ class Validator {
       this.nodeKey,
-      '-lruntime=debug',
+      ...newCliArgs,
+      '-lruntime=debug,gateway=debug,pallet_cash=debug,ethereum_client=debug',
```
Do we want all of these to be `debug`?

hayesgm commented on scripts/build_release.sh:
```diff
@@ -15,41 +15,53 @@ echo "*** Building release $version ***"
 wasm="./target/release/wbuild/gateway-runtime/gateway_runtime.compact.wasm"
 
-if [ ! -f "$bin" -o ! -f "$wasm" -o ! -f "$types" -o ! -f "$rpc" -o ! -f "$contracts" ]; then
+if [ ! -f "$bin" -o ! -f "$wasm" -o ! -f "$types" -o ! -f "$rpc" -o ! -f "$contracts" -o ]; then
```
Dangling `-o`

hayesgm commented on scripts/build_release.sh:
```diff
@@ -15,41 +15,53 @@ echo "*** Building release $version ***"
+echo "*** Building testnet release gateway ***"
 
-cargo +nightly build --release
+WASM_BUILD_RUSTFLAGS='--cfg feature="testnet"' cargo +nightly build --release --features testnet
```
Is this now testnet? Should it be flagged as such?

hayesgm commented on scripts/build_release.sh:
```diff
@@ -15,41 +15,53 @@ echo "*** Building release $version ***"
 echo "Built release $version"
-echo "  wasm: $release_dir/gateway_runtime.compact.wasm"
-echo "  wasm.checksum: $release_dir/gateway_runtime.checksum"
+echo "  wasm: $release_dir/gateway-testnet.wasm"
```
This probably _shouldn't_ be tagged test-net now?

waynenilsen commented on node/src/chain_spec.rs:
```diff
@@ -99,6 +89,10 @@ fn development_genesis() -> GenesisConfig {
             .expect("cannot get system time for genesis")
             .as_millis() as Timestamp,
+        // Starports
+        vec![],
```
This is causing `cargo run -- --dev --tmp --alice` to fail with panic at `pallets/cash/src/lib.rs:713`

```
Thread 'main' panicked at 'Starports must be set in the genesis config',
```

we should probably change something here .... i can look into what we were doing before

hayesgm replied:
cool, though maybe it should fail since we need a starport to get going, right?

waynenilsen commented:
I also just noticed that in `build_spec.js` the starport address is still being set under the `properties` attribute

```

async function setStarport(chainSpec, chainConfig, opts) {
  let chainDeployment = await fetchChainDeployment(chainConfig.eth_network);
  let starportAddress = must(chainDeployment.Contracts, 'Starport');
  // TODO: Set other properties
  chainSpec.properties.eth_starport_address = starportAddress;
}
```

should be modified

hayesgm committed:
Minor m10 Tweaks (#338)

* Minor m10 Tweaks

This just patches a few small topics for m10. Nothing serious in the scheme of things.

* Add changelog and fix jest

* Fix upgrade to m9 then to m10 test

jflatow committed:
Fix chain cash regression (do not modify on transfer) (#337)

* Fix chain cash regression (do not modify on transfer)

* Rebase Geoff's unit test fixes

* Set Starports should go to genesis not properties, and others fixes

hayesgm committed:
Fix feature
